ssl/ssl_key_share.cc

```diff
@@ -294,6 +295,233 @@ class CECPQ2KeyShare : public SSLKeyShare {
   HRSS_private_key hrss_private_key_;
 };
 
+class KEMKeyShare : public SSLKeyShare {
+ public:
+  KEMKeyShare(int nid, uint16_t group_id) : nid_(nid), group_id_(group_id) {}
+
+  uint16_t GroupID() const override { return group_id_; }
+
+  bool Offer(CBB *out) override {
+    // Ensure |out| is valid, has no children, and has been initialized.
+    // We check that |out| has no children because otherwise CBB_data()
+    // will produce a fatal error by way of assert(cbb->child == NULL).
+    if (!out || out->child || !CBB_data(out)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should not have been initialized (by e.g. a previous call
```

Comment:
why not? The EVP API is designed such that a single `EVP_PKEY_CTX` can be used for multiple key generations

Reply:
That is true for the EVP APIs, but I argue should not be the case for `SSLKeyShare` objects. This was my way of enforcing that a single `SSLKeyShare` object should be used only to generate/negotiate a single key share. Even if the case of an HRR or renegotiation, the handshake code should create a new `SSLKeyShare` object instead of reusing an existing one (this is the way the code is written at present). My reasoning for this is just based on code hygiene and preventing bugs from popping up in the future. The way things are written currently, it _should_ be safe to reuse a key share object, but since the key shares maintain some sort state, I could imagine situations where the code flow gets changed and bugs occur because the key share is trying to Accept() or Finish() based on a previously stored, incorrect private key.

Reply:
> this is the way the code is written at present

which code?

Reply:
right now nothing stops the user to call Finish multiple times with the same key share, but Offer and Accept can be called only once. Should we destroy the state at the end of Finish to ensure it can't be called more than once?

Reply:
> > this is the way the code is written at present
> 
> which code?

The [client key share/handshake code](https://github.com/aws/aws-lc/blob/d1c1bfcf015e88fd022f7cfcc205d9e9f2c3412b/ssl/extensions.cc#L2291-L2330). This code path is used for both the initial ClientHello, and a ClientHello sent as in response to an HRR. Point being that a new key share is always `Create()`ed.